pygmt/helpers/tempfile.py

```diff
@@ -126,18 +126,27 @@ def tempfile_from_geojson(geojson):
         E.g. '1a2b3c4d5e6.gmt'.
     """
     with GMTTempFile(suffix=".gmt") as tmpfile:
+        # pylint: disable=import-outside-toplevel
+        import geopandas as gpd
+
         os.remove(tmpfile.name)  # ensure file is deleted first
         ogrgmt_kwargs = {"filename": tmpfile.name, "driver": "OGR_GMT", "mode": "w"}
         try:
+            # Workaround to map int/int64 to int32
```

Comment:
Need to add a unit test to `pygmt/tests/test_geopandas.py` that uses int/int64 columns. Probably a good idea to check uint/uint64 too.

Reply:
> Probably a good idea to check uint/uint64 too.

See line https://github.com/geopandas/geopandas/blob/35463adda663dfa57c9809865ebbb15448823278/geopandas/io/file.py#L640
```
out_type = type(np.zeros(1, in_type).item()).__name__
```
For any integer-like data types (like int, int32, int64, unit, uint64 and more), the `out_type` is always `int`.

Reply:
Ah ok, then a unit test with just int64 columns should be fine, no need for parametrized tests.

Reply:
@weiji14 I have never used geopandas and I find it challenging for me to add a geopandas test for this PR. Could you please help with it?

Reply:
Getting a little late for me, but let me post the test example I'm working on:

```python
import geopandas as gpd
import pandas as pd
from pygmt import Figure, info, which, makecpt

shapefile = which(
    fname="@RidgeTest.shp @RidgeTest.shx @RidgeTest.dbf @RidgeTest.prj",
    download="c",
)
gdf = gpd.read_file(shapefile[0])
gdf["geometry"] = gdf.to_crs(crs="EPSG:3857").buffer(distance=100000)
gdf["NPOINTS"] = gdf.NPOINTS.astype(dtype="int32")
gdf.plot()

fig = Figure()
makecpt(cmap="lajolla", series=[10, 60, 10], continuous=True)
fig.plot(
    data=gdf[["NPOINTS", "geometry"]],
    projection="X5c",
    frame=True,
    pen="1p,black",
    close=True,
    fill="+z",
    cmap=True,
    aspatial="Z=NPOINTS",
)
fig.colorbar()
fig.show()
```

Couldn't get it to plot though, getting some weird errors like:

```
plot [ERROR]: Bad OGR/GMT: @n not allowed before FEATURE_DATA
plot [ERROR]: Aspatial associations set with -a but input file is not in OGR/GMT format!
```
Might be another bug? I'll try to come up with an alternative example tomorrow.

Reply:
[pygmt-2m6nvbqu.gmt.txt](https://github.com/GenericMappingTools/pygmt/files/12318310/pygmt-2m6nvbqu.gmt.txt)

This is the temporary OGR/GMT file generated by `tempfile_from_geojson` (with an extra `.txt` suffix so that I can upload it to GitHub).

The first few lines are:
```
# @VGMT1.0 @GPOLYGON
# @R-4506555.399/-3052902.89568/3804018.88695/5612191.52675               
# @Je3857
# @Jp"+proj=merc +a=6378137 +b=6378137 +lat_ts=0 +lon_0=0 +x_0=0 +y_0=0 +k=1 +units=m +nadgrids=@null +wktext +no_defs"
# @Jw"PROJCS[\"WGS 84 / Pseudo-Mercator\",GEOGCS[\"WGS 84\",DATUM[\"WGS_1984\",SPHEROID[\"WGS 84\",6378137,298.257223563,AUTHORITY[\"EPSG\",\"7030\"]],AUTHORITY[\"EPSG\",\"6326\"]],PRIMEM[\"Greenwich\",0,AUTHORITY[\"EPSG\",\"8901\"]],UNIT[\"degree\",0.0174532925199433,AUTHORITY[\"EPSG\",\"9122\"]],AUTHORITY[\"EPSG\",\"4326\"]],PROJECTION[\"Mercator_1SP\"],PARAMETER[\"central_meridian\",0],PARAMETER[\"scale_factor\",1],PARAMETER[\"false_easting\",0],PARAMETER[\"false_northing\",0],UNIT[\"metre\",1],AXIS[\"Easting\",EAST],AXIS[\"Northing\",NORTH],EXTENSION[\"PROJ4\",\"+proj=merc +a=6378137 +b=6378137 +lat_ts=0 +lon_0=0 +x_0=0 +y_0=0 +k=1 +units=m +nadgrids=@null +wktext +no_defs\"],AUTHORITY[\"EPSG\",\"3857\"]]"
# @NNPOINTS
# @Tstring
# FEATURE_DATA
>
```
Apparently, GMT incorrectly parse the `@null` string in the `@Jp` and `@Jw` lines as an invalid OGR/GMT code. So it looks like an upstream bug.

Reply:
Oo, didn't notice that `@null`. I see at https://proj.org/en/9.2/usage/transformation.html#the-null-grid that you can specify null, by why would there be an `@` symbol?!! A little bit of searching in the https://github.com/OSGeo/PROJ repo shows that this might be valid though?

Do we need to report this to GMT, geopandas or PROJ? In the meantime, I can try a projection that doesn't have `@null` in the PROJ string.

Reply:
> I see at https://proj.org/en/9.2/usage/transformation.html#the-null-grid that you can specify null, by why would there be an `@` symbol?!!

I think `@` means the grid is optional (see https://github.com/OSGeo/PROJ/blob/0372c81511a11afa76e01ebb30adb00bcb950772/NEWS#L2936). Also see the search results of `=@`: https://github.com/search?q=repo%3AOSGeo%2FPROJ+%3D%40&type=code

Reply:
Ok, I've added the unit test at commit c1b3b437b2cddc72861c0e6d9064c9c83cc67c1f, and used a projection that doesn't have `@null`, feel free to modify if needed. Also see that you've opened https://github.com/GenericMappingTools/gmt/issues/7712 to track the `@null` bug.

Reply:
> [pygmt-2m6nvbqu.gmt.txt](https://github.com/GenericMappingTools/pygmt/files/12318310/pygmt-2m6nvbqu.gmt.txt)
> 
> This is the temporary OGR/GMT file generated by `tempfile_from_geojson` (with an extra `.txt` suffix so that I can upload it to GitHub).
> 
> The first few lines are:
> 
> ```
> # @VGMT1.0 @GPOLYGON
> # @R-4506555.399/-3052902.89568/3804018.88695/5612191.52675               
> # @Je3857
> # @Jp"+proj=merc +a=6378137 +b=6378137 +lat_ts=0 +lon_0=0 +x_0=0 +y_0=0 +k=1 +units=m +nadgrids=@null +wktext +no_defs"
> # @Jw"PROJCS[\"WGS 84 / Pseudo-Mercator\",GEOGCS[\"WGS 84\",DATUM[\"WGS_1984\",SPHEROID[\"WGS 84\",6378137,298.257223563,AUTHORITY[\"EPSG\",\"7030\"]],AUTHORITY[\"EPSG\",\"6326\"]],PRIMEM[\"Greenwich\",0,AUTHORITY[\"EPSG\",\"8901\"]],UNIT[\"degree\",0.0174532925199433,AUTHORITY[\"EPSG\",\"9122\"]],AUTHORITY[\"EPSG\",\"4326\"]],PROJECTION[\"Mercator_1SP\"],PARAMETER[\"central_meridian\",0],PARAMETER[\"scale_factor\",1],PARAMETER[\"false_easting\",0],PARAMETER[\"false_northing\",0],UNIT[\"metre\",1],AXIS[\"Easting\",EAST],AXIS[\"Northing\",NORTH],EXTENSION[\"PROJ4\",\"+proj=merc +a=6378137 +b=6378137 +lat_ts=0 +lon_0=0 +x_0=0 +y_0=0 +k=1 +units=m +nadgrids=@null +wktext +no_defs\"],AUTHORITY[\"EPSG\",\"3857\"]]"
> # @NNPOINTS
> # @Tstring
> # FEATURE_DATA
> >
> ```
> 
> Apparently, GMT incorrectly parse the `@null` string in the `@Jp` and `@Jw` lines as an invalid OGR/GMT code. So it looks like an upstream bug.
> 
> Edit: opened a bug report in [GenericMappingTools/gmt#7712](https://github.com/GenericMappingTools/gmt/issues/7712).

FYI, the upstream bug has been fixed.